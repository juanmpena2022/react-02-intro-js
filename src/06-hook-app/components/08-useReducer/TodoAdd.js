import { useForm } from "../../../hooks/06-hook-app/useForm";
import PropTypes from "prop-types";

export const TodoAdd = ({ handleAddTodo }) => {
  const {
    formState: { description },
    handleInputChange,
    reset,
  } = useForm({
    description: "",
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length === 0) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Add Todo</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          className="form-control"
          name="description"
          value={description}
          placeholder="Aprender..."
          autoComplete="off"
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Agregar
        </button>
      </form>
    </>
  );
};

TodoAdd.propTypes = {
  handleAddTodo: PropTypes.func.isRequired,
};
