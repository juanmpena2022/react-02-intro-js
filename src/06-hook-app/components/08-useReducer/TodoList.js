import PropTypes from "prop-types";
import { TodoListItem } from "./TodoListItem";

export const TodoList = ({ todos = [], handleDelete, handleToggle }) => {
  return (
    <>
      <ul className="list-group list-group-flush">
        {todos.map((t, i) => (
          <TodoListItem
            key={t.id}
            todo={t}
            index={i}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        ))}
      </ul>
    </>
  );
};

TodoList.propTypes = {
  todos: PropTypes.array,
  handleDelete: PropTypes.func.isRequired,
  handleToggle: PropTypes.func.isRequired,
};
