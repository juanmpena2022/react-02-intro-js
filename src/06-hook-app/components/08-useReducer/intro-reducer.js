//1. Estado inicial
const initialState = [
  {
    id: 1,
    todo: "Comprar Pan",
    done: false,
  },
];

//2. Funcion Reducer
const todoReducer = (state = initialState, action) => {
  if (action?.type === "add") {
    return [...state, action.payload];
  }
  return state;
};

//3. Obtener state por defecto
let todos = todoReducer();

//4. Nuevo valor a adicionar
const newTodo = {
  id: 2,
  todo: "Comprar Leche",
  done: false,
};

//5. Accion que representa adicionar
const actionAdd = {
  type: "add",
  payload: newTodo,
};

//6. Obtener state con accionde adicionar
todos = todoReducer(todos, actionAdd);

console.log(todos);
