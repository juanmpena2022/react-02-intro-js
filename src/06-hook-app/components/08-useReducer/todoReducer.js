import { todoReducerActionType } from "../../../enums/todoReducerActionType";

export const todoReducer = (state = [], action) => {
  switch (action?.type) {
    case todoReducerActionType.add:
      return [...state, action.payload];
    case todoReducerActionType.delete:
      return state.filter((s) => s.id !== action.payload);
    case todoReducerActionType.toggle:
      return state.map((s) => {
        if (s.id === action.payload) {
          s.done = !s.done;
        }
        return s;
      });
    default:
      return state;
  }
};
