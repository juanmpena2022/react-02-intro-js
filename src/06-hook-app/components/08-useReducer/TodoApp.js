import { useEffect, useReducer } from "react";
import { todoReducerActionType } from "../../../enums/todoReducerActionType";
import "./style.css";
import { TodoAdd } from "./TodoAdd";
import { TodoList } from "./TodoList";
import { todoReducer } from "./todoReducer";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
  const [todos, todosDispatch] = useReducer(todoReducer, [], init);
  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (id) => {
    const actionDelete = {
      type: todoReducerActionType.delete,
      payload: id,
    };
    todosDispatch(actionDelete);
  };
  const handleToggle = (id) => {
    const actionToggle = {
      type: todoReducerActionType.toggle,
      payload: id,
    };
    todosDispatch(actionToggle);
  };

  const handleAddTodo = (todo) => {
    todosDispatch({
      type: todoReducerActionType.add,
      payload: todo,
    });
  };

  return (
    <>
      <h1>TodoApp: ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </>
  );
};
