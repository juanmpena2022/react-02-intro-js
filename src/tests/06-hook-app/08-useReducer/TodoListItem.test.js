import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import { TodoListItem } from "../../../06-hook-app/components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";

describe("pruebas del componente <TodoListItem />", () => {
  const handleToggle = jest.fn();
  const handleDelete = jest.fn();

  test("snapshot", () => {
    jest.clearAllMocks();
    const index = 0;
    const todo = demoTodos[index];
    const wrapper = shallow(
      <TodoListItem
        todo={todo}
        index={index}
        handleToggle={handleToggle}
        handleDelete={handleDelete}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("debe llamar la funcion de borrar", () => {
    jest.clearAllMocks();
    const index = 0;
    const todo = demoTodos[index];
    const wrapper = shallow(
      <TodoListItem
        todo={todo}
        index={index}
        handleToggle={handleToggle}
        handleDelete={handleDelete}
      />
    );
    const btn = wrapper.find('button');
    btn.simulate('click');
    expect(handleDelete).toHaveBeenCalledWith(todo.id);
  });

  test("debe llamar la funcion toggle", () => {
    jest.clearAllMocks();
    const index = 0;
    const todo = demoTodos[index];
    const wrapper = shallow(
      <TodoListItem
        todo={todo}
        index={index}
        handleToggle={handleToggle}
        handleDelete={handleDelete}
      />
    );
    const p = wrapper.find('p');
    p.simulate('click');
    expect(handleToggle).toHaveBeenCalledWith(todo.id);
  });

  test("debe mostrar el valor correctamente", () => {
    const index = 0;
    const todo = demoTodos[index];
    const wrapper = shallow(
      <TodoListItem
        todo={todo}
        index={index}
        handleToggle={handleToggle}
        handleDelete={handleDelete}
      />
    );
    const p = wrapper.find('p');
    expect(p.text()).toBe(`${index + 1}. ${todo.desc}`);
  });
});
