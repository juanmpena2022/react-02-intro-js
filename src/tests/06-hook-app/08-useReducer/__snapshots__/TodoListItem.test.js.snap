// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas del componente <TodoListItem /> snapshot 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false text-center"
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
